pkg/app/piped/apistore/eventstore/store.go

```diff
@@ -0,0 +1,154 @@
+	latestEventMap atomic.Value
+}
+
+const (
+	defaultSyncInterval = time.Minute
+)
+
+// NewStore creates a new event store instance.
+// This syncs with the control plane to keep the list of events for this runner up-to-date.
+func NewStore(apiClient apiClient, gracePeriod time.Duration, logger *zap.Logger) Store {
+	s := &store{
+		apiClient:    apiClient,
+		syncInterval: defaultSyncInterval,
+		gracePeriod:  gracePeriod,
+		logger:       logger.Named("event-store"),
+	}
+	s.latestEventMap.Store(make(map[string]*model.Event))
+	return s
+}
+
+// Run starts runner that periodically makes the Events in the cache up-to-date
+// by fetching from the control-plane.
+func (s *store) Run(ctx context.Context) error {
+	s.logger.Info("start running event store")
+
+	syncTicker := time.NewTicker(s.syncInterval)
+	defer syncTicker.Stop()
+
+	// Do first sync without waiting the first ticker.
+	s.milestone = time.Now().Add(-time.Hour).Unix()
+	s.sync(ctx)
+
+	for {
+		select {
+		case <-syncTicker.C:
+			s.sync(ctx)
+
+		case <-ctx.Done():
+			s.logger.Info("event store has been stopped")
+			return nil
+		}
+	}
+}
+
+// sync fetches a list of events inside the range between own milestone and the current local time.
+// Only this function takes responsibility for updating the cache.
+func (s *store) sync(ctx context.Context) error {
+	// TODO: Use UTC and let control-plane convert it into the control-plane's Local time
+	// Unexpected behavior can be happened if Piped uses different timezone from the control-plane.
+	to := time.Now().Unix()
+	resp, err := s.apiClient.ListEvents(ctx, &pipedservice.ListEventsRequest{
+		From:  s.milestone,
+		To:    to,
+		Order: pipedservice.ListEventsOrder_FROM_OLDEST,
+	})
+	if err != nil {
+		s.logger.Error("failed to list events", zap.Error(err))
+		return err
+	}
+	// Set this time's "to" as the next time's "from".
+	s.milestone = to
+
+	// Make the cache up-to-date by traversing events sorted by oldest first.
+	eventMap := s.latestEventMap.Load().(map[string]*model.Event)
+	for _, event := range resp.Events {
+		id := model.EventDefinitionID(event.Name, event.Labels)
+		eventMap[id] = event
+	}
+	s.latestEventMap.Store(eventMap)
+
+	return nil
+}
+
+func (s *store) Getter() Getter {
+	return s
+}
+
+func (s *store) GetLatest(ctx context.Context, name string, labels map[string]string) (*model.Event, bool) {
+	eventMap := s.latestEventMap.Load().(map[string]*model.Event)
+	id := model.EventDefinitionID(name, labels)
+	event, ok := eventMap[id]
+	if ok {
+		return event, true
+	}
+
+	// If not found in the cache, fetch from the control-plane.
+	resp, err := s.apiClient.GetLatestEvent(ctx, &pipedservice.GetLatestEventRequest{
+		Name:   name,
+		Labels: labels,
+	})
+	if err != nil {
+		s.logger.Error("failed to get the latest event", zap.Error(err))
+		return nil, false
+	}
+
+	// NOTE: Don't update the cache to prevent it from being overwritten by slightly older ones.
```

Comment:
Why we don't update the list?
The event has `CreatedAt` field, so we can use that to check whether the event should be updated or not.

And I think in the microservices list of our users, there are always some applications that have not been actively developed compared to others. Their events will be very sparse. So without saving into the list, piped will always make the call to control-plane to get the same one.

Reply:
The `sync` function periodically updates the cache, and it works in another goroutine. What I'm most worried about is, while processing `GetLatest`, `sync` can update the cache. I mean:

```go
	// NOTE: Don't update the cache to prevent it from being overwritten by slightly older ones.
	eventMap = s.latestEventMap.Load().(map[string]*model.Event)
	eventMap[id] = resp.Event        // I'm most worried about the "sync()" function updates the cache at this point.
	s.latestEventMap.Store(eventMap) // Then the latest cache is overwritten by the old one here.
```

Reply:
If Piped was run on the multi-core processor, I think it can happen, but let me know if I'm wrong.

Reply:
The data race can be resolved by some ways:
- change from using atomic to using a mutex. (fetches the list from control-plane, lock, and update). I think it is fine enough for our case: 1 call/m and EventWatcher.
- use 2 separate maps, one for LIST API, one for GET API

Reply:
Ah, not fancy 😄  The first one looks enough! Only two goroutines can access this and it rarely happens to wait for unlocking. Thank you for telling me!